metricflow/query/group_by_item/candidate_push_down/push_down_visitor.py

```diff
@@ -0,0 +1,408 @@
+                        parent_issues=(),
+                    )
+                ),
+            )
+        return PushDownResult(
+            candidate_set=intersected_candidate_set,
+            issue_set=merged_issue_set,
+        )
+
+    @override
+    def visit_metric_node(self, node: MetricGroupByItemResolutionNode) -> PushDownResult:
+        """At the metric node, intersect candidates from the parents and pass them to the children.
+
+        This node can represent a metric that prevents querying by some types of group-by items. To model that
+        restriction, this filters the appropriate candidates.
+        """
+        with self._path_from_start_node_tracker.track_node_visit(node) as current_traversal_path:
+            merged_result_from_parents = self._merge_push_down_results_from_parents(
+                push_down_results_from_parents={
+                    parent_node: parent_node.accept(self) for parent_node in node.parent_nodes
+                },
+                current_traversal_path=current_traversal_path,
+            )
+            logger.info(
+                f"Handling {node.ui_description} with candidates from parents:\n"
+                + indent_log_line(mf_pformat(merged_result_from_parents.candidate_set.specs))
+            )
+            if merged_result_from_parents.candidate_set.is_empty:
+                return merged_result_from_parents
+
+            metric = self._semantic_manifest_lookup.metric_lookup.get_metric(node.metric_reference)
+
+            # For metrics with offset_to_grain, don't allow date_part group-by-items
+            patterns_to_apply: Sequence[SpecPattern] = ()
+            if metric.type is MetricType.SIMPLE or metric.type is MetricType.CUMULATIVE:
+                pass
+            elif metric.type is MetricType.RATIO or metric.type is MetricType.DERIVED:
+                for input_metric in metric.input_metrics:
+                    if input_metric.offset_to_grain:
+                        # From comment in previous query parser:
+                        # "Cannot extract date part for metrics with offset_to_grain."
+                        patterns_to_apply = (NoneDatePartPattern(),)
+                        break
+            else:
+                assert_values_exhausted(metric.type)
+
+            candidate_specs: Sequence[LinkableInstanceSpec] = merged_result_from_parents.candidate_set.specs
+            issue_sets_to_merge = [merged_result_from_parents.issue_set]
+
+            if len(patterns_to_apply) == 0:
+                return PushDownResult(
+                    candidate_set=GroupByItemCandidateSet(
+                        specs=tuple(candidate_specs),
+                        measure_paths=merged_result_from_parents.candidate_set.measure_paths,
+                        path_from_leaf_node=current_traversal_path,
+                    ),
+                    issue_set=MetricFlowQueryResolutionIssueSet.merge_iterable(issue_sets_to_merge),
+                )
```

Comment:
This is fine here, but we don't actually need it, right?

The cumulative metrics logic in the measure visitor doesn't include the early return, so we'll get those "after applying patterns" log entries for all measures but not for all metrics. Does that matter?

Reply:
It was originally here to cut down on the log output, but I think it's better to remove it for consistency in log output.